pandas/_libs/tslibs/src/datetime/pd_datetime.c

```diff
@@ -0,0 +1,2356 @@
+                         "in datetime string \"%s\"",
+                         str);
+          }
+          goto error;
+        }
+      } else if (sublen >= 1 && isdigit(substr[0])) {
+        offset_minute = substr[0] - '0';
+        ++substr;
+        --sublen;
+      } else {
+        goto parse_error;
+      }
+    }
+
+    /* Apply the time zone offset */
+    if (offset_neg) {
+      offset_hour = -offset_hour;
+      offset_minute = -offset_minute;
+    }
+    if (out_local != NULL) {
+      *out_local = 1;
+      // Unlike NumPy, do not change internal value to local time
+      *out_tzoffset = 60 * offset_hour + offset_minute;
+    }
+  }
+
+  /* Skip trailing whitespace */
+  while (sublen > 0 && isspace(*substr)) {
+    ++substr;
+    --sublen;
+    comparison =
+        compare_format(&format, &format_len, " ", 1, format_requirement);
+    if (comparison == COMPARISON_ERROR) {
+      goto parse_error;
+    } else if (comparison == COMPLETED_PARTIAL_MATCH) {
+      goto finish;
+    }
+  }
+
+  if ((sublen != 0) || (format_len != 0)) {
+    goto parse_error;
+  }
+
+finish:
+  if (out_bestunit != NULL) {
+    *out_bestunit = bestunit;
+  }
+  return 0;
+
+parse_error:
+  if (want_exc) {
+    PyErr_Format(PyExc_ValueError,
+                 "Error parsing datetime string \"%s\" at position %d", str,
+                 (int)(substr - str));
+  }
+  return -1;
+
+error:
+  return -1;
+}
+
+/* Converts the int64_t representation of a duration to ISO; mutates len */
+static char *int64ToIsoDuration(int64_t value, size_t *len) {
+  pandas_timedeltastruct tds;
+  int ret_code;
+
+  pandas_timedelta_to_timedeltastruct(value, NPY_FR_ns, &tds);
+
+  // Max theoretical length of ISO Duration with 64 bit day
+  // as the largest unit is 70 characters + 1 for a null terminator
+  char *result = PyObject_Malloc(71);
+  if (result == NULL) {
+    PyErr_NoMemory();
+    return NULL;
+  }
+
+  ret_code = make_iso_8601_timedelta(&tds, result, len);
+  if (ret_code == -1) {
+    PyErr_SetString(PyExc_ValueError,
+                    "Could not convert timedelta value to string");
+    PyObject_Free(result);
+    return NULL;
+  }
+
+  return result;
+}
+
+/*
+ * This function returns a pointer to the DateTimeMetaData
+ * contained within the provided datetime dtype.
+ *
+ * Copied near-verbatim from numpy/core/src/multiarray/datetime.c
+ */
+static PyArray_DatetimeMetaData
+get_datetime_metadata_from_dtype(PyArray_Descr *dtype) {
+  return (((PyArray_DatetimeDTypeMetaData *)dtype->c_metadata)->meta);
+}
+
+static void pandas_datetime_destructor(PyObject *op) {
+  void *ptr = PyCapsule_GetPointer(op, PandasDateTime_CAPSULE_NAME);
+  PyMem_Free(ptr);
+}
+
+static int pandas_datetime_exec(PyObject *module) {
+  PyDateTime_IMPORT;
+  PandasDateTime_CAPI *capi = PyMem_Malloc(sizeof(PandasDateTime_CAPI));
+  if (capi == NULL) {
+    PyErr_NoMemory();
+    return -1;
+  }
+  capi->npy_datetimestruct_to_datetime = npy_datetimestruct_to_datetime;
```

Comment:
The macros in the header file dispatch to these struct members, which in turn reference static functions defined within this module